Userland/Libraries/LibTLS/Certificate.cpp

```diff
@@ -10,536 +11,849 @@
+
+            EXIT_SCOPE();
+        }
+    }
+
+    if (!decoder.eof()) {
+        ERROR_WITH_SCOPE("Reached end of TBS parse with more data left"sv);
+    }
+
+    certificate.is_self_issued = TRY(certificate.issuer.to_string()) == TRY(certificate.subject.to_string());
 
-    dbgln_if(TLS_DEBUG, "Certificate issued for {} by {}", certificate.subject.subject, certificate.issuer.subject);
+    EXIT_SCOPE();
 
     return certificate;
+}
 
-#undef DROP_OBJECT_OR_FAIL
-#undef ENSURE_OBJECT_KIND
-#undef ENTER_SCOPE_OR_FAIL
-#undef ENTER_SCOPE_WITHOUT_TYPECHECK
-#undef EXIT_SCOPE
-#undef READ_OBJECT_OR_FAIL
+ErrorOr<Certificate> Certificate::parse_certificate(ReadonlyBytes buffer, bool)
+{
+    Crypto::ASN1::Decoder decoder { buffer };
+    Vector<StringView, 8> current_scope {};
+
+    // Certificate ::= SIGNED{TBSCertificate}
+
+    // SIGNED{ToBeSigned} ::= SEQUENCE {
+    //     toBeSigned ToBeSigned,
+    //     COMPONENTS OF SIGNATURE{ToBeSigned},
+    // }
+
+    // SIGNATURE{ToBeSigned} ::= SEQUENCE {
+    //      algorithmIdentifier AlgorithmIdentifier{{SupportedAlgorithms}},
+    //      encrypted ENCRYPTED-HASH{ToBeSigned},
+    // }
+
+    // ENCRYPTED-HASH{ToBeSigned} ::= BIT STRING (CONSTRAINED BY {
+    // -- shall be the result of applying a hashing procedure to the DER-encoded (see 6.2)
+    // -- octets of a value of -- ToBeSigned -- and then applying an encipherment procedure
+    // -- to those octets -- } )
+
+    ENTER_TYPED_SCOPE(Sequence, "Certificate"sv);
+
+    Certificate certificate = TRY(parse_tbs_certificate(decoder, current_scope));
+    certificate.original_asn1 = TRY(ByteBuffer::copy(buffer));
+
+    CertificateKeyAlgorithm signature_algorithm = TRY(parse_algorithm_identifier(decoder, current_scope));
+    certificate.signature_algorithm = signature_algorithm;
+
+    PUSH_SCOPE("signature"sv);
+    READ_OBJECT(BitString, Crypto::ASN1::BitStringView, signature);
+    certificate.signature_value = TRY(ByteBuffer::copy(signature.raw_bytes()));
+    POP_SCOPE();
+
+    if (!decoder.eof()) {
+        ERROR_WITH_SCOPE("Reached end of Certificate parse with more data left"sv);
+    }
+
+    EXIT_SCOPE();
+
+    return certificate;
 }
 
+#undef PUSH_SCOPE
+#undef ENTER_SCOPE
+#undef ENTER_TYPED_SCOPE
+#undef POP_SCOPE
+#undef EXIT_SCOPE
+#undef READ_OBJECT
+#undef DROP_OBJECT
+#undef REWRITE_TAG
+
+ErrorOr<String> RelativeDistinguishedName::to_string()
+{
+#define ADD_IF_RECOGNIZED(identifier, shorthand_code)                 \
+    do {                                                              \
+        if (it->key == identifier) {                                  \
+            cert_name.appendff("\\{}={}", shorthand_code, it->value); \
+            continue;                                                 \
+        }                                                             \
+    } while (0);
+
+    StringBuilder cert_name;
+
+    for (auto it = m_members.begin(); it != m_members.end(); ++it) {
```

Comment:
Not sure if we want to include the duplicated values? (`MAIL` and `E`)

Also couldnt find the OID for `host name` and `ip` in https://www.iana.org/assignments/ldap-parameters/ldap-parameters.xhtml#ldap-parameters-3 so not sure what it would be..